dap-mode.el

```diff
@@ -683,14 +693,37 @@ thread exection but the server will log message."
   (run-hook-with-args 'dap-terminated-hook debug-session)
   (dap--refresh-breakpoints))
 
+(defun dap--output-buffer-format-with-category (category output)
+  "Formats a string suitable for printing to the output buffer using CATEGORY and OUTPUT."
+  (let ((message (format "%s: %s" category output)))
+    (if (string= (substring message -1) "\n")
+        message
+      (concat message "\n"))))
+
+(defun dap--output-buffer-format (output-body)
+  "Formats a string suitable for printing to the output buffer using an OUTPUT-BODY."
+  (if dap-label-output-buffer-category
+      (dap--output-buffer-format-with-category (gethash "category" output-body) (gethash "output" output-body))
+    (gethash "output" output-body)))
+
+(defun dap--print-to-output-buffer (debug-session str)
+  "Insert content from STR into the output buffer associated with DEBUG-SESSION."
+  (with-current-buffer (dap--debug-session-output-buffer debug-session)
+    (let ((win (get-buffer-window (current-buffer))))
+      (goto-char (point-max))
+      (insert str)
+      (set-window-point win (point-max)))))
+
 (defun dap--on-event (debug-session event)
   "Dispatch EVENT for DEBUG-SESSION."
   (let ((event-type (gethash "event" event)))
     (pcase event-type
-      ("output" (with-current-buffer (dap--debug-session-output-buffer debug-session)
-                  (save-excursion
-                    (goto-char (point-max))
-                    (insert (gethash "output" (gethash "body" event))))))
+      ("output" (let* ((event-body (gethash "body" event))
+                       (formatted-output (dap--output-buffer-format event-body)))
+                  (if dap-output-buffer-filter
```

Comment:
nit: seems like you could join the `if` and `when` via or, eg `(when (or (not dap-ouput-buffer-filter) (member...`

Reply:
Good point! 

I had originally had it condensed down to the solution you proposed, but I thought it was less understandable in terms of intent, and decided to go with readability over conciseness because of this.

Although, if that's OK with you, I'm happy to change it back!